backend/services/workflows/app/worker/worker_group.go

```diff
@@ -117,7 +120,93 @@ func (w *workerGroup) RunWorker(ctx context.Context) {
 	l.Info("worker starting up")
 	// workerSidecar is responsible for notifying the broker about slow workflows
 	go w.workerSidecar(ctx, sidecarChan, sidecarDone)
-	w.workerMain(ctx, sidecarChan, sidecarDone)
+	switch w.sub.Type() {
+	case natsio.PullSubscription:
+		w.pullWorkerMain(ctx, sidecarChan, sidecarDone)
+	case natsio.ChanSubscription:
+		w.workerMain(ctx, sidecarChan, sidecarDone)
+	default:
+		l.Error("unsupported consumer type")
+	}
+}
+
+func (w *workerGroup) pullWorkerMain(
+	ctx context.Context,
+	sidecarChan chan *natsio.Msg,
+	sidecarDone chan struct{},
+) {
+	l := log.FromContext(ctx)
+	ctxDone := ctx.Done()
+	timeoutTimer := newStoppedTimer()
+	for {
+		var msg *natsio.Msg
+		select {
+		case <-ctx.Done():
+			return
+		case <-sidecarDone:
+			return
+		default:
+			msgs, err := w.sub.Fetch(1, natsio.MaxWait(1*time.Second))
+			if err != nil {
+				if errors.Is(err, natsio.ErrTimeout) {
+					continue
+				}
+				l.Errorf("failed to fetch message: %s", err)
+				return
+			}
+			if len(msgs) == 0 {
+				continue
+			}
+			msg = msgs[0]
+		}
+
+		// Notify the sidecar routine about the new message
+		select {
+		case sidecarChan <- msg:
+
+		case <-timeoutTimer.After(w.notifyPeriod / 8):
+			l.Warn("timeout notifying sidecar routine about message")
+
+		case <-sidecarDone:
+			return
+		case <-ctxDone:
+			return
+		}
+
+		job := &model.Job{}
+		err := json.Unmarshal(msg.Data, job)
+		if err != nil {
+			l.Error(errors.Wrap(err, "failed to unmarshall message"))
+			if err := msg.Term(); err != nil {
+				l.Error(errors.Wrap(err, "failed to term the message"))
+			}
+			continue
+		}
+		// process the job
+		l.Infof("processing job %s workflow %s", job.ID, job.WorkflowName)
+		err = processJob(ctx, job, w.store, w.client)
+		if err != nil {
+			l.Errorf("error processing job: %s", err.Error())
+		} else {
+			l.Infof("finished job %s workflow %s", job.ID, job.WorkflowName)
+		}
+		// stop the in progress ticker and ack the message
+		select {
+		case sidecarChan <- nil:
+
+		case <-timeoutTimer.After(w.notifyPeriod):
+			l.Errorf("timeout notifying sidecar about job completion")
+
+		case <-ctxDone:
+			return
+		case <-sidecarDone:
+			return
+		}
+		// Release message
+		if err := msg.AckSync(); err != nil {
+			l.Error(errors.Wrap(err, "failed to ack the message"))
+		}
```

Comment:
There's quite a bit of code duplication here. Maybe you could refactor and only wrap the first `select` branch in a push/pull `func`?